content/tutorials/use-specialized-payment-types/create-custom-transactors.md

```diff
@@ -0,0 +1,379 @@
+
+    explicit CreateCheck(ApplyContext& ctx) : Transactor(ctx)
+    {
+    }
+
+    static NotTEC
+    preflight(PreflightContext const& ctx);
+
+    static TER
+    preclaim(PreclaimContext const& ctx);
+
+    TER
+    doApply() override;
+};
+
+}  // namespace ripple
+```
+
+Initializing the transactor with `ApplyContext` gives it access to:
+
+- The transaction that triggered the transactor.
+- A view of the SLE.
+- A journal to log errors.
+
+
+## CPP File
+
+### 1. Add a `preflight` function.
+
+The `preflight` function checks for errors before accessing the ledger and incurring fees.
+
+- `PreflightContext` doesn't have a view of the ledger.
+- Use bracket notation to retrieve fields from ledgers and transactions:
+        
+        auto const curExpiration = (*sle*)[~sfExpiration];
+        (*sle)[sfBalance] = (*sle)[sfBalance] + reqDelta;
+            
+    **Note:** The `~` symbol returns an optional type.
+
+- You can view ledger and transaction schemas here:
+    - [`LedgerFormats.cpp`](https://github.com/XRPLF/rippled/blob/master/src/ripple/protocol/impl/LedgerFormats.cpp)
+    - [`TxFormats.cpp`](https://github.com/XRPLF/rippled/blob/master/src/ripple/protocol/impl/TxFormats.cpp)
+
+
+```c++
+CreateCheck::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureChecks))
+        return temDISABLED;
+
+    NotTEC const ret{preflight1(ctx)};
+    if (!isTesSuccess(ret))
+        return ret;
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+    {
+        // There are no flags (other than universal) for CreateCheck yet.
+        JLOG(ctx.j.warn()) << "Malformed transaction: Invalid flags set.";
+        return temINVALID_FLAG;
+    }
+    if (ctx.tx[sfAccount] == ctx.tx[sfDestination])
+    {
+        // They wrote a check to themselves.
+        JLOG(ctx.j.warn()) << "Malformed transaction: Check to self.";
+        return temREDUNDANT;
+    }
+
+    {
+        STAmount const sendMax{ctx.tx.getFieldAmount(sfSendMax)};
+        if (!isLegalNet(sendMax) || sendMax.signum() <= 0)
+        {
+            JLOG(ctx.j.warn()) << "Malformed transaction: bad sendMax amount: "
+                            << sendMax.getFullText();
+            return temBAD_AMOUNT;
+        }
+
+        if (badCurrency() == sendMax.getCurrency())
+        {
+            JLOG(ctx.j.warn()) << "Malformed transaction: Bad currency.";
+            return temBAD_CURRENCY;
+        }
+    }
+
+    if (auto const optExpiry = ctx.tx[~sfExpiration])
+    {
+        if (*optExpiry == 0)
+        {
+            JLOG(ctx.j.warn()) << "Malformed transaction: bad expiration";
+            return temBAD_EXPIRATION;
+        }
+    }
+
+    return preflight2(ctx);
+}
+```
+
+
+### 2. Add a `preclaim` function.
+
+The `preclaim` functions checks for errors that require viewing information on the ledger.
+
+- `PreclaimContext` has a read-only view of the ledger.
+- This function incurs fees.
```

Comment:
What fees? How? Who is paying? Please clarify. I suspect the intent is that a transaction that gets to this stage generally gets put into a ledger and burns the transaction cost even if it fails (so, the sender of the transaction pays). Of course, that's not strictly true if they used `fail_hard`, but I'm not sure how that comes in.

The biggest distinction/turning point where a transaction _should_ burn the transaction cost if at all possible is when it's relayed to other servers. I believe that if a transaction gets to `preclaim` then it's probably going to be relayed unless using `fail_hard` but I'm not totally sure. 

I also think the class of result code used may have an effect on that behavior, but I'm not confident about the details. But looking at the sample code, it looks like `preclaim` should return either `tesSUCCESS` (if all went well so far) or a `tec` code  (if something is wrong). Maybe you aren't supposed to return other classes of codes (like `ter` or `tef`) in a `preclaim` function?